controllers/datadoghq/tls.go

```diff
@@ -0,0 +1,232 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package datadoghq
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/DataDog/watermarkpodautoscaler/apis/datadoghq/v1alpha1"
+)
+
+const (
+	certificateCacheTimeout           = 10 * time.Minute
+	certificateCacheExpirationTimeout = 10 * time.Minute
+	certificateErrorCacheTimeout      = 1 * time.Minute
+)
+
+type tlsTransport struct {
+	wrappedTransport *http.Transport
+}
+
+// tlsCertificateCache is an expiring cache to store certificates in memory
+// used for TLS connections
+type tlsCertificateCache struct {
+	mu    sync.RWMutex
+	cache map[string]*tlsCacheEntry
+}
+
+// tlsCacheEntry is an entry of the certificate cache
+type tlsCacheEntry struct {
```

Comment:
The types and method types are all mixed. For readability, it'd be better to have:
```
type

type methods

type

type methods
```

Free function at the end except `new` right after `type`